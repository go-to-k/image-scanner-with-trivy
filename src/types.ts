/**
 * Output configuration for scan logs to CloudWatch Logs.
 */
export interface CloudWatchLogsOutput {
  readonly type: 'cloudWatchLogs';
  readonly logGroupName: string;
}

/**
 * Tagged union type of output configurations for scan logs.
 */
export type ScanLogsOutputOptions = CloudWatchLogsOutput;

/**
 * Lambda function event object for Scanner Custom Resource.
 */
export interface ScannerCustomResourceProps {
  readonly addr: string;
  readonly imageUri: string;
  readonly ignoreUnfixed: string;
  readonly severity: string[];
  readonly scanners: string[];
  readonly imageConfigScanners: string[];
  readonly exitCode: number;
  readonly exitOnEol: number;
  readonly trivyIgnore: string[];
  readonly platform: string;
  readonly output?: ScanLogsOutputOptions;
}
