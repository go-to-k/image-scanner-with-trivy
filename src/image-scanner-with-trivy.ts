import { join } from 'path';
import { CustomResource, Duration, Size, Token } from 'aws-cdk-lib';
import { IRepository } from 'aws-cdk-lib/aws-ecr';
import { Platform } from 'aws-cdk-lib/aws-ecr-assets';
import {
  Architecture,
  AssetCode,
  Handler,
  Runtime,
  SingletonFunction,
} from 'aws-cdk-lib/aws-lambda';
import { ILogGroup } from 'aws-cdk-lib/aws-logs';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

/**
 * Enum for Severity Selection
 *
 * @see https://aquasecurity.github.io/trivy/latest/docs/scanner/vulnerability/#severity-selection
 */
export enum Severity {
  UNKNOWN = 'UNKNOWN',
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL',
}

/**
 * Enum for Scanners
 *
 * @see https://aquasecurity.github.io/trivy/latest/docs/configuration/others/#enabledisable-scanners
 */
export enum Scanners {
  VULN = 'vuln',
  CONFIG = 'config',
  SECRET = 'secret',
  LICENSE = 'license',
}

/**
 * Enum for ImageConfigScanners
 *
 * @see https://aquasecurity.github.io/trivy/latest/docs/target/container_image/#container-image-metadata
 */
export enum ImageConfigScanners {
  CONFIG = 'config',
  SECRET = 'secret',
}

/**
 * Represents the output of the scan logs.
 */
export abstract class ScanLogsOutput {
  /**
   * Scan logs output to CloudWatch Logs log group.
   */
  public static cloudWatchLogs(logGroup: ILogGroup): ScanLogsOutput {
    return {
      type: 'cloudWatchLogs',
      logGroupName: logGroup.logGroupName,
    };
  }

  /**
   * The type of the scan logs output.
   */
  public abstract readonly type: 'cloudWatchLogs';

  /**
   * The name of the logGroup.
   */
  public abstract readonly logGroupName?: string;
}

export interface ImageScannerWithTrivyProps {
  /**
   * Image URI for scan target.
   */
  readonly imageUri: string;

  /**
   * Repository including the image URI for scan target.
   *
   * Because of grantPull to CustomResourceLambda.
   */
  readonly repository: IRepository;

  /**
   * The unfixed/unfixable vulnerabilities mean that the patch has not yet been provided on their distribution.
   *
   * To hide unfixed/unfixable vulnerabilities, you can use the `--ignore-unfixed` flag.
   *
   * @default false
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/scanner/vulnerability/#unfixed-vulnerabilities
   */
  readonly ignoreUnfixed?: boolean;

  /**
   * Severity Selection
   *
   * The severity is taken from the selected data source since the severity from vendors is more accurate.
   * Using CVE-2023-0464 as an example, while it is rated as "HIGH" in NVD, Red Hat has marked its 'Impact' as "Low". As a result, Trivy will display it as "Low".
   *
   * The severity depends on the compile option, the default configuration, etc. NVD doesn't know how the vendor distributes the software.
   * Red Hat evaluates the severity more accurately. That's why Trivy prefers vendor scores over NVD.
   *
   * It defaults to `CRITICAL` IN THIS CONSTRUCT for safety in CI/CD, but the default configuration of Trivy is "CRITICAL,HIGH,MEDIUM,LOW,UNKNOWN".
   *
   * @default [Severity.CRITICAL]
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/scanner/vulnerability/#severity-selection
   */
  readonly severity?: Severity[];

  /**
   * Enable/Disable Scanners
   *
   * You can enable/disable scanners with the `scanners`.
   *
   * For example, container image scanning enables vulnerability (VULN) and secret scanners (SECRET) by default.
   * If you don't need secret scanning, it can be disabled by specifying Scanners.VULN only.
   *
   * @default [Security.VULN,Scanners.SECRET]
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/configuration/others/#enabledisable-scanners
   */
  readonly scanners?: Scanners[];

  /**
   * Enum for ImageConfigScanners
   *
   * Container images have configuration. docker inspect and `docker history` show the information according to the configuration.
   * Trivy scans the configuration of container images for
   *
   * - Misconfigurations
   * - Secrets
   *
   * They are disabled by default. You can enable them with `imageConfigScanners`.
   *
   * @default []
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/target/container_image/#container-image-metadata
   */
  readonly imageConfigScanners?: ImageConfigScanners[];

  /**
   * Exit Code
   *
   * Use the `exitCode` option if you want to exit with a non-zero exit code.
   *
   * You can specify 0 if you do not want to exit even when vulnerabilities are detected.
   *
   * It defaults to 1 IN THIS CONSTRUCT for safety in CI/CD. In the original trivy, it is 0.
   *
   * @default 1
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/configuration/others/#exit-code
   */
  readonly exitCode?: number;

  /**
   * Exit on EOL
   *
   * Sometimes you may surprisingly get 0 vulnerabilities in an old image:
   *  - Enabling --ignore-unfixed option while all packages have no fixed versions.
   *  - Scanning a rather outdated OS (e.g. Ubuntu 10.04).
   *
   * An OS at the end of service/life (EOL) usually gets into this situation, which is definitely full of vulnerabilities.
   * `exitOnEol` can fail scanning on EOL OS with a non-zero code.
   *
   * It defaults to 1 IN THIS CONSTRUCT for safety in CI/CD. In the original trivy, it is 0.
   *
   * @default 1
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/configuration/others/#exit-on-eol
   */
  readonly exitOnEol?: number;

  /**
   * By Finding IDs
   *
   * The ignore rules written to the .trivyignore in trivy.
   * Put each line you write in the file into one element of the array.
   *
   * @example
   *     $ cat .trivyignore
   *     # Accept the risk
   *     CVE-2018-14618
   *
   *     # Accept the risk until 2023-01-01
   *     CVE-2019-14697 exp:2023-01-01
   *
   *     # No impact in our settings
   *     CVE-2019-1543
   *
   *     # Ignore misconfigurations
   *     AVD-DS-0002
   *
   *     # Ignore secrets
   *     generic-unwanted-rule
   *     aws-account-id
   *
   * @default []
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/configuration/filtering/#trivyignore
   */
  readonly trivyIgnore?: string[];

  /**
   * Memory Size (MB) for Scanner Lambda
   *
   * You can specify between `3008` and `10240`.
   *
   * If this Construct execution terminates abnormally due to SIGKILL, try a larger size.
   *
   * Default value (`3008` MB) is Maximum Lambda memory size for default AWS account without quota limit increase.
   *
   * @default 3008
   */
  readonly memorySize?: number;

  /**
   * Scan Image on a specific Architecture and OS
   *
   * By default, Trivy loads an image on a `linux/amd64` machine.
   *
   * To customize this, pass a `platform` argument in the format OS/Architecture for the image, such as `linux/arm64`
   *
   * @default -
   */
  readonly platform?: string;

  /**
   * Configuration for scan logs output
   *
   * By default, scan logs are output to default log group created by Scanner Lambda.
   *
   * Specify this if you want to send scan logs to other than the default log group.
   *
   * Currently, only `cloudWatchLogs` is supported.
   *
   * @default - scan logs output to default log group created by Scanner Lambda(`/aws/lambda/${functionName}`)
   */
  readonly scanLogsOutput?: ScanLogsOutput;
}

// Maximum Lambda memory size for default AWS account without quota limit increase
const DEFAULT_MEMORY_SIZE = 3008;

export class ImageScannerWithTrivy extends Construct {
  constructor(scope: Construct, id: string, props: ImageScannerWithTrivyProps) {
    super(scope, id);

    if (
      props.memorySize &&
      !Token.isUnresolved(props.memorySize) &&
      (props.memorySize < 3008 || props.memorySize > 10240)
    ) {
      throw new Error(
        `You can specify between \`3008\` and \`10240\` for \`memorySize\`, got ${props.memorySize}.`,
      );
    }

    const customResourceLambda = new SingletonFunction(this, 'CustomResourceLambda', {
      uuid: '470b6343-d267-f753-226c-1e99f09f319a',
      lambdaPurpose: 'Custom::ImageScannerWithTrivyCustomResourceLambda',
      runtime: Runtime.FROM_IMAGE,
      handler: Handler.FROM_IMAGE,
      code: AssetCode.fromAssetImage(join(__dirname, '../assets/lambda'), {
        platform: Platform.LINUX_ARM64,
      }),
      architecture: Architecture.ARM_64,
      timeout: Duration.seconds(900),
      retryAttempts: 0,
      memorySize: props.memorySize ?? DEFAULT_MEMORY_SIZE,
      ephemeralStorageSize: Size.gibibytes(10), // for cases that need to update trivy DB: /tmp/trivy/db/trivy.db
    });
    props.repository.grantPull(customResourceLambda);

    const imageScannerProvider = new Provider(this, 'Provider', {
      onEventHandler: customResourceLambda,
    });

    const imageScannerProperties: {
      [key: string]: string | string[] | boolean | number | ScanLogsOutput;
    } = {};
    imageScannerProperties.addr = this.node.addr;
    imageScannerProperties.imageUri = props.imageUri;
    imageScannerProperties.ignoreUnfixed = props.ignoreUnfixed ?? false;
    imageScannerProperties.severity = props.severity ?? [Severity.CRITICAL];
    imageScannerProperties.scanners = props.scanners ?? [];
    imageScannerProperties.imageConfigScanners = props.imageConfigScanners ?? [];
    imageScannerProperties.exitCode = props.exitCode ?? 1;
    imageScannerProperties.exitOnEol = props.exitOnEol ?? 1;
    imageScannerProperties.trivyIgnore = props.trivyIgnore ?? [];
    imageScannerProperties.platform = props.platform ?? '';

    if (props.scanLogsOutput) {
      imageScannerProperties.output = props.scanLogsOutput;
    }

    new CustomResource(this, 'Resource', {
      resourceType: 'Custom::ImageScannerWithTrivy',
      properties: imageScannerProperties,
      serviceToken: imageScannerProvider.serviceToken,
    });
  }
}
