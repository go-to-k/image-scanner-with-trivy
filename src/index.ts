import { join } from 'path';
import { CustomResource, Duration } from 'aws-cdk-lib';
import { IRepository } from 'aws-cdk-lib/aws-ecr';
import {
  Architecture,
  AssetCode,
  Handler,
  Runtime,
  SingletonFunction,
} from 'aws-cdk-lib/aws-lambda';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

/**
 * Enum for Severity Selection
 *
 * @see https://aquasecurity.github.io/trivy/v0.45/docs/scanner/vulnerability/#severity-selection
 */
export enum Severity {
  UNKNOWN = 'UNKNOWN',
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL',
}

export interface ImageScannerWithTrivyProps {
  /**
   * Image URI for scan target.
   */
  readonly imageUri: string;

  /**
   * Repository including the image URI for scan target.
   *
   * Because of grantPull to CustomResourceLambda.
   */
  readonly repository: IRepository;

  /**
   * The unfixed/unfixable vulnerabilities mean that the patch has not yet been provided on their distribution.
   *
   * To hide unfixed/unfixable vulnerabilities, you can use the `--ignore-unfixed` flag.
   *
   * @see https://aquasecurity.github.io/trivy/latest/docs/scanner/vulnerability/#unfixed-vulnerabilities
   */
  readonly ignoreUnfixed?: boolean;

  /**
   * Severity Selection
   *
   * The severity is taken from the selected data source since the severity from vendors is more accurate.
   * Using CVE-2023-0464 as an example, while it is rated as "HIGH" in NVD, Red Hat has marked its 'Impact' as "Low". As a result, Trivy will display it as "Low".
   *
   * The severity depends on the compile option, the default configuration, etc. NVD doesn't know how the vendor distributes the software.
   * Red Hat evaluates the severity more accurately. That's why Trivy prefers vendor scores over NVD.
   *
   * @default [] - The default configuration of Trivy is "CRITICAL,HIGH,MEDIUM,LOW,UNKNOWN".
   *
   * @see https://aquasecurity.github.io/trivy/v0.45/docs/scanner/vulnerability/#severity-selection
   */
  readonly severity?: Severity[];
}

export class ImageScannerWithTrivy extends Construct {
  constructor(scope: Construct, id: string, props: ImageScannerWithTrivyProps) {
    super(scope, id);

    const { imageUri, repository, ignoreUnfixed, severity } = props;

    const customResourceLambda = new SingletonFunction(this, 'CustomResourceLambda', {
      uuid: '470b6343-d267-f753-226c-1e99f09f319a',
      lambdaPurpose: 'Custom::ImageScannerWithTrivyCustomResourceLambda',
      runtime: Runtime.FROM_IMAGE,
      handler: Handler.FROM_IMAGE,
      code: AssetCode.fromAssetImage(join(__dirname, '../assets/lambda')),
      architecture: Architecture.ARM_64,
      timeout: Duration.seconds(900),
      retryAttempts: 0,
      memorySize: 3008, // Maximum memory size for Default AWS account
    });
    repository.grantPull(customResourceLambda);

    const imageScannerProvider = new Provider(this, 'Provider', {
      onEventHandler: customResourceLambda,
    });

    const imageScannerProperties: { [key: string]: string | string[] | boolean } = {};
    imageScannerProperties.addr = this.node.addr;
    imageScannerProperties.imageUri = imageUri;
    imageScannerProperties.ignoreUnfixed = ignoreUnfixed ?? false; // TODO: boolean or string ?
    imageScannerProperties.severity = severity ?? [];

    new CustomResource(this, 'Default', {
      resourceType: 'Custom::ImageScannerWithTrivy',
      properties: imageScannerProperties,
      serviceToken: imageScannerProvider.serviceToken,
    });
  }
}
