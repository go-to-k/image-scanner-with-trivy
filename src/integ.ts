import { resolve } from 'path';
import { App, Stack } from 'aws-cdk-lib';
import { DockerImageAsset, Platform } from 'aws-cdk-lib/aws-ecr-assets';
import { ImageScannerWithTrivy, Scanners, Severity } from '.';

const app = new App();
const stack = new Stack(app, 'ImageScannerWithTrivyStack');

const image = new DockerImageAsset(stack, 'DockerImage', {
  directory: resolve(__dirname, '../assets/lambda'),
  platform: Platform.LINUX_ARM64,
});

new ImageScannerWithTrivy(stack, 'ImageScannerWithTrivy4', {
  imageUri: image.imageUri,
  repository: image.repository,
  ignoreUnfixed: true,
  severity: [Severity.CRITICAL],
  scanners: [Scanners.VULN, Scanners.SECRET],
  exitCode: 1,
  exitOnEol: 1,
});
