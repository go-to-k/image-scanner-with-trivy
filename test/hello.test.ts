import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { ImageScannerWithTrivy, Scanners, Severity } from '../src';

const getTemplate = (): Template => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const repository = new Repository(stack, 'ImageRepository', {});

  new ImageScannerWithTrivy(stack, 'ImageScannerWithTrivy', {
    imageUri: 'imageUri',
    repository: repository,
    ignoreUnfixed: true,
    severity: [Severity.CRITICAL, Severity.HIGH],
    scanners: [Scanners.VULN, Scanners.SECRET],
    exitCode: 1,
    exitOnEol: 1,
  });
  return Template.fromStack(stack);
};

describe('Fine-grained Assertions Tests', () => {
  const template = getTemplate();

  test('Snapshot test', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('ImageScannerWithTrivy created', () => {
    template.resourceCountIs('Custom::ImageScannerWithTrivy', 1);
  });
});
