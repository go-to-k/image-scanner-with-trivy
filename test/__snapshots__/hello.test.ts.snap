// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fine-grained Assertions Tests Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319a884A34E3": {
      "DependsOn": [
        "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRoleDefaultPolicyB1A38321",
        "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRole676D3F17",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ImageUri": {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:81545a70461d1e74eba9366a56f1d2759ed6b9017cb074e0468495a88ec64690",
          },
        },
        "MemorySize": 3008,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRole676D3F17",
            "Arn",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aEventInvokeConfig2DEF6CCE": {
      "Properties": {
        "FunctionName": {
          "Ref": "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319a884A34E3",
        },
        "MaximumRetryAttempts": 0,
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRole676D3F17": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRoleDefaultPolicyB1A38321": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ImageRepositoryBBCBC9DF",
                  "Arn",
                ],
              },
            },
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRoleDefaultPolicyB1A38321",
        "Roles": [
          {
            "Ref": "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319aServiceRole676D3F17",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ImageRepositoryBBCBC9DF": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
    "ImageScannerWithTrivy": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "ImageScannerWithTrivyProviderframeworkonEvent7AD0141E",
            "Arn",
          ],
        },
        "addr": "c86a2a03c1252a9e84323d4e55c74690e11e4a44ce",
        "exitCode": 1,
        "exitOnEol": 1,
        "ignoreUnfixed": true,
        "imageConfigScanners": [],
        "imageUri": "imageUri",
        "scanners": [
          "vuln",
          "secret",
        ],
        "severity": [
          "CRITICAL",
          "HIGH",
        ],
        "trivyIgnore": [
          "CVE-2023-37920",
          "CVE-2019-14697 exp:2023-01-01",
        ],
      },
      "Type": "Custom::ImageScannerWithTrivy",
      "UpdateReplacePolicy": "Delete",
    },
    "ImageScannerWithTrivyProviderframeworkonEvent7AD0141E": {
      "DependsOn": [
        "ImageScannerWithTrivyProviderframeworkonEventServiceRoleDefaultPolicy1013358F",
        "ImageScannerWithTrivyProviderframeworkonEventServiceRoleA9C12442",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/ImageScannerWithTrivy/Provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319a884A34E3",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "ImageScannerWithTrivyProviderframeworkonEventServiceRoleA9C12442",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ImageScannerWithTrivyProviderframeworkonEventServiceRoleA9C12442": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ImageScannerWithTrivyProviderframeworkonEventServiceRoleDefaultPolicy1013358F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319a884A34E3",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CustomImageScannerWithTrivyCustomResourceLambda470b6343d267f753226c1e99f09f319a884A34E3",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ImageScannerWithTrivyProviderframeworkonEventServiceRoleDefaultPolicy1013358F",
        "Roles": [
          {
            "Ref": "ImageScannerWithTrivyProviderframeworkonEventServiceRoleA9C12442",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
